data/man/man1/booh.1
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program; if not, write to the Free Software
.\" Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
.\"

.TH BOOH 1 "August 2008" "BOOH" "The Booh web-album"

.SH NAME
Booh \- a static web-album generator

.SH SYNOPSIS
\fBbooh\fR [\fIOPTION\fR]... [\fIFILENAME\fR]

.SH DESCRIPTION
\fBBooh\fR is a static Web-Album generator. It's a program which
takes one or several series of photos and videos, and
automatically build static web pages to browse them, creating
thumbnails etc.
.PP
If you invoke \fBbooh\fR from the command-line, this will launch
the graphical user interface, which is the most user friendly way
of creating an album, editing etc. If you append a
\fIFILENAME\fR, \fBbooh\fR will try to open this file when
starting up.
.PP
Notice that prior to using \fBbooh\fR to create a Web-Album, your
photos and videos must be classified; if they are not, you may
use the dedicated tool \fBbooh-classifier\fR.

.SH OPTIONS
.TP
.BR -h , \ --help
show command-line options summary
.TP
\fB-v \fINUM\fR, \fB--verbose \fINUM\fR
Set max verbosity level (0: errors, 1: warnings, 2: important messages, 3: other messages)

.SH AUTHOR
Written by Guillaume Cottenceau.
.br
Visit official homepage: http://booh.org/

.SH SEE ALSO
.BR booh-classifier(1)
.BR booh-backend(1)

.SH COPYRIGHT
Copyright \(co 2005-2008 Guillaume Cottenceau.
.br
This is Free Software; this software is licensed under the GPL version 2, as published by the Free Software Foundation.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.


ChangeLog
[33mcommit b1f899a5074bf9c23433cd3d6f6a49e739efb8f9[m
Author: Guillaume Cottenceau <gcottenc@gmail.com>
Date:   Sat Nov 2 21:26:54 2013 +0100

    themes demos with 0.9.4

[33mcommit de58bf561d075d13fc5bfc117dae0fefcb751541[m
Author: Guillaume Cottenceau <gcottenc@gmail.com>
Date:   Sat Nov 2 21:18:15 2013 +0100

    bump copyright to 2013

[33mcommit 6b41b211ecce258c4d9fce7696855f1552c85c91[m
Author: Guillaume Cottenceau <gcottenc@gmail.com>
Date:   Sat Nov 2 21:12:20 2013 +0100

    make "gradient" the default theme

[33mcommit b9005a644bab5b8ea162443a463c57371bedc8d4[m
Author: Guillaume Cottenceau <gcottenc@gmail.com>
Date:   Sat Nov 2 21:10:26 2013 +0100

    make only one size selected by default

[33mcommit 0bf4c978a63d0645608eea8458860514a1ce4a0f[m
Author: Guillaume Cottenceau <gcottenc@gmail.com>
Date:   Sat Nov 2 21:10:13 2013 +0100

    make "gradient" the default theme

[33mcommit 5b5b6c49356c03908d2e4c9ea679c2a7938db04f[m
Author: Guillaume Cottenceau <gcottenc@gmail.com>
Date:   Sat Nov 2 21:03:07 2013 +0100

    0.9.4

[33mcommit 760cb3e978dd0e30affcf0d1700666ff4d061de6[m
Author: Guillaume Cottenceau <gcottenc@gmail.com>
Date:   Sat Nov 2 20:59:20 2013 +0100

    README redirects to website

[33mcommit aa5966557701016a05a15b4b15f8fbc37d753432[m
Author: Guillaume Cottenceau <gcottenc@gmail.com>
Date:   Sat Nov 2 20:58:00 2013 +0100

    update to flowplayer 3.2.16

[33mcommit 4b15812329111657a2c9c436551cbc12eec207c2[m
Author: Guillaume Cottenceau <gcottenc@gmail.com>
Date:   Thu Aug 1 21:51:17 2013 +0200

    fix starting broken when multiple cpus and no existing config

[33mcommit 8065958c9f08ee0a5483bcd8364ea7f8f1895fdc[m
Author: Guillaume Cottenceau <gcottenc@gmail.com>
Date:   Mon Jul 8 23:17:01 2013 +0200

    drop support for 800x600 screens and add support for 1920x1080

[33mcommit 9216e52f1185d737df11739b0d191f1e311019fe[m
Author: Guillaume Cottenceau <gcottenc@gmail.com>
Date:   Sat Sep 15 22:32:45 2012 +0200

    this is not perl!

[33mcommit 0aa89b2ae8253cda1baf0299a91cb190dd4624a9[m
Author: Guillaume Cottenceau <gcottenc@gmail.com>
Date:   Thu Aug 30 16:18:55 2012 +0200

    slight simplification

[33mcommit 0dfcebbd01a54524a0b8745fe22c4e7422fe02a4[m
Author: Guillaume Cottenceau <gcottenc@gmail.com>
Date:   Thu Aug 30 16:18:41 2012 +0200

    do not include www in tarball!

[33mcommit 90af7ca8aabcec527caa5c630674cebf0df70317[m
Author: Guillaume Cottenceau <gcottenc@gmail.com>
Date:   Sat Aug 25 15:58:07 2012 +0200

    fix small stupid bug

[33mcommit 308a2029925d463211f03243d5ff6dc7dbb65ba7[m
Author: Guillaume Cottenceau <gcottenc@gmail.com>
Date:   Wed Aug 1 21:55:57 2012 +0200

    prevent from "regular expression too big" in gettext due to stupid untranslatable anyway debug string

[33mcommit 90b7c7805f2b956c596fed5e3faea3ba701b1acd[m
Author: Guillaume Cottenceau <gcottenc@gmail.com>
Date:   Tue May 29 22:51:30 2012 +0200

    slightly reworked lftp connect error message

[33mcommit 3814f5d163d8be5ab9ae713d777d63ffffc20f51[m
Author: Guillaume Cottenceau <gcottenc@gmail.com>
Date:   Sun May 20 22:01:08 2012 +0200

    make all web pages smartphone compatible

[33mcommit c0df3e0506012e0fe48650c1998131b1344fd355[m
Merge: 35b01b4 a58973c
Author: Guillaume Cottenceau <gcottenc@gmail.com>
Date:   Sat May 19 22:21:53 2012 +0200

    Merge branch 'master' of ssh://git.zarb.org/srv/git/booh

[33mcommit 35b01b40e74aafb045ced2b71c7f1269f246749c[m
Author: Guillaume Cottenceau <gcottenc@gmail.com>
Date:   Sat May 19 22:21:43 2012 +0200

    in case suddenly more CPUs are available, talk about appropriate configuration (upgrading to new hardware..)

[33mcommit a58973ce8f747f17b6f32e221cde0d96c01c7b39[m
Merge: 1e3f3e6 a1a9347
Author: Guillaume Cottenceau <gcottenc@gmail.com>
Date:   Mon Mar 12 17:42:22 2012 +0100

    Merge branch 'master' of ssh://zarb.org/srv/git/booh

[33mcommit 1e3f3e658c2a5ec6c8d30fea1051816ddce1b92c[m
Author: Benoit Dien <benoit.dien@gmail.com>
Date:   Sun Aug 7 01:30:20 2011 +0200

    Page speed: Include the CSS before any javascript to allow parallel loading

[33mcommit dadc3737acd00a938f033f4267ecb8a8e68abaa3[m
Author: Benoit Dien <benoit.dien@gmail.com>
Date:   Sun Aug 7 01:20:52 2011 +0200

    Enable Apache caching (30min)

[33mcommit a2732e47a922ef0961794a281d61e315ef422085[m
Author: Benoit Dien <benoit.dien@gmail.com>
Date:   Sun Aug 7 01:13:00 2011 +0200

    Page speed: specify image size

[33mcommit 4cde67ed86e57d7acba08468b40609d0fe6ec312[m
Author: Benoit Dien <benoit.dien@gmail.com>
Date:   Sun Aug 7 00:53:43 2011 +0200

    Optimize PNG files

[33mcommit a1a9347e9c6c6579fd2ccef1cd40cf102bf4d949[m
Author: Guillaume Cottenceau <gcottenc@gmail.com>
Date:   Mon Mar 5 11:23:14 2012 +0100

    preliminary work on making it somehow correct on handsets

[33mcommit ee5034646d32eef4baec8ec3f8e95cec10f9a9f9[m
Author: Guillaume Cottenceau <gcottenc@gmail.com>
Date:   Mon Feb 27 21:51:03 2012 +0100

    save website

[33mcommit 89bafd44d12c4b5ca803f5c3e894e879681a27a8[m
Author: Guillaume Cottenceau <gcottenc@gmail.com>
Date:   Sat Feb 18 14:11:20 2012 +0100

    also rescue Gtk::InitError for booh-backend

[33mcommit 93a9bb39802339e42df62810e2a84f955314fbaa[m
Author: Benoit Dien <benoit.dien@gmail.com>
Date:   Thu Aug 4 11:09:27 2011 +0200

    Tweak shadows CSS More subtle shadows (Thanks Amaury) Remove "-webkit-box-shadow", webkit can handle the "box-shadow" property now

[33mcommit 6a9c028542490f3fa17c3ac0c1d3810096238090[m
Author: Benoit Dien <benoit.dien@gmail.com>
Date:   Fri Jul 29 23:20:44 2011 +0200

    Dim the shadow (Change transparency)

[33mcommit 9d6863378fa254655001f9370cadc67d0b48e5a8[m
Author: Benoit Dien <benoit.dien@gmail.com>
Date:   Sat Jul 30 23:33:27 2011 +0200

    Add the image size to keep formatting even when the image is still loading

[33mcommit ebe4d1a56578bbdbf9f66ee39972f69448d315f6[m
Author: Benoit Dien <benoit.dien@gmail.com>

data/booh/themes/sbs/root/booh.js
var images_ary = new Array();
var images_loaded = new Array();
var current = 0;
var slideshow = 0;
var slideshow_pause = null;
var slideshow_timer = null;

for (i = 0; i < images.length; i++) { 
    /* this array will contain 0 if image not yet loaded, 1 when loading,
     * 2 when complete */
    images_loaded[i] = 0;
}

function dbg(t) {
    document.getElementById('dbg_text').innerHTML += t + "\n";
}

/* load image, return 1 if image is finished loading */
function load(i) {
    if (images_loaded[i] == 0) {
        images_ary[i] = new Image();
        images_ary[i].src = images[i];
        images_loaded[i] = 1;
    }
    if (images_loaded[i] == 1) {
        if (images_ary[i].complete) {
            images_loaded[i] = 2;
        } else {
            return 0;
        }
    }
    return 1;
}

function getparam(key) {
    all_params = location.href.split("#");
    if (all_params.length > 1) {
        params = all_params[1].split("&");
        for (i = 0; i < params.length; i++) {
            keyvalue = params[i].split("=");
            if (keyvalue[0] == key) {
                return keyvalue[1];
            }
        }
    }
    return null;
}

function loadcurrent(img) {
    for (i = 0; i < images.length; i++) {
        if (images[i] == img) {
            current = i;
            display_current();
            return;
        }
    }
    current = 0;
    display_current();
}

function browser_href() {
    all = location.href.split("/");
    return all[all.length - 1];
}

/* check URL for changes; allows the URL to reflect currently showed image */
var currentURL = '';
function checkURL() {
    if (currentURL == 'ignore1') {
        // do nothing
    } else if (currentURL == 'ignore2') {
        currentURL = browser_href();
    } else {
        href = browser_href();
        if (href != currentURL) {
            currentURL = href;
            img = getparam('current');
            loadcurrent(img);
        }
    }
    setTimeout("checkURL()", 50);
}

function preload() { 

    /* favor current image, if user clicked on `last' or something */
    load(current);

    /* don't blindly preload all images at the beginning,
     * but rather load them one by one, in order to get
     * next ones faster, beginning with next to current
     */
    if (current + 1 < images.length && load(current + 1) == 0) {
        setTimeout("preload()", 50);
        return;
    }
    if (current - 1 >= 0 && load(current - 1) == 0) {
        setTimeout("preload()", 50);
        return;
    }

    for (i = current + 2; i < images.length && i <= current + 5; i++) { 
        if (load(i) == 0) {
            setTimeout("preload()", 50);
            return;
        }
    }
    for (i = current - 2; i >= current - 3; i--) { 
        if (i >= 0) {
            if (load(i) == 0) {
                setTimeout("preload()", 50);
                return;
            }
        }
    }

    setTimeout("preload()", 50);
}

function add_cookie(val) {
    var expires = new Date(new Date().getTime() + (30 * 86400000));  // 30 days
    document.cookie = val
                      + '; expires=' + expires.toGMTString()
                      + '; path=/';
}

function get_cookie(key) {
    if (document.cookie) {
        var index = document.cookie.indexOf(key);
        if (index != -1) {
            var oleft = (document.cookie.indexOf('=', index) + 1);
            var oright = document.cookie.indexOf(';', index);
            if (oright == -1) {
                oright = document.cookie.length;
            }
            return document.cookie.substring(oleft, oright);
        }
    }
    return null;
}

function init() {

    preferred_pause = get_cookie('booh-slideshow-pause-' + dbltilda_theme);
    if (preferred_pause != null) {
        document.getElementById('secs').value = preferred_pause;
    }

    if (getparam('run_slideshow')) {
        toggle_slideshow();
    }

    checkURL();

    preload();

    if (images.length == 1) {
        document.getElementById("b_slideshow").disabled = true;
        document.getElementById("b_slideshow").setAttribute("class", "disabled");
    }

    if (navigator.userAgent.indexOf('Opera') == -1) {
        document.onkeydown = keyDownEvent;
    }
}

function update_sensibilities() {
    if (current == 0) {
        // prevent focus to stay on disabled buttons - then keyboard navigation won't work
        document.getElementById('thumbnails').focus();
        document.getElementById("b_first").disabled = true;
        document.getElementById("b_previous").disabled = true;
        document.getElementById("b_first").setAttribute("class", "disabled");
        document.getElementById("b_previous").setAttribute("class", "disabled");
    } else {
        document.getElementById("b_first").disabled = false;
        document.getElementById("b_previous").disabled = false;
        document.getElementById("b_first").removeAttribute("class");
        document.getElementById("b_previous").removeAttribute("class");
    }

    if (current == images.length - 1) {
        // prevent focus to stay on disabled buttons - then keyboard navigation won't work
        document.getElementById('thumbnails').focus();
        document.getElementById("b_next").disabled = true;
        document.getElementById("b_last").disabled = true;
        document.getElementById("b_next").setAttribute("class", "disabled");
        document.getElementById("b_last").setAttribute("class", "disabled");
    } else {
        document.getElementById("b_next").disabled = false;
        document.getElementById("b_last").disabled = false;
        document.getElementById("b_next").removeAttribute("class");
        document.getElementById("b_last").removeAttribute("class");
    }
}

function set_cursor_(value, element) {

    if (!element || !element.style) {
        return;
    }

    element.style.cursor = value;

    children = element.childNodes;
    for (i = 0; i < children.length; i++) {
        set_cursor_(value, children.item[i]);
    }
}

function set_cursor(value) {
    set_cursor_(value, document.getElementById('body'));
    set_cursor_(value, document.getElementById('b_first'));
    set_cursor_(value, document.getElementById('b_previous'));
    set_cursor_(value, document.getElementById('b_next'));
    set_cursor_(value, document.getElementById('b_last'));
}

function show_current_text() {
    /* don't show text if image not yet loaded because navigator
     * won't refresh it during load */
    if (images_loaded[current] == 2) {
        document.getElementById('image_counter').innerHTML = ( current + 1 ) + "/" + images.length;
        document.getElementById('main_text').innerHTML = captions[current];
        for (i = 0; i < other_sizes.length; i++) { 
            if (other_sizes[i] == "original") {
                var original = eval("elements_" + other_sizes[i] + "[current]");
                if (original != undefined) {
                    document.getElementById('link' + other_sizes[i]).href = original;
                    document.getElementById('link' + other_sizes[i]).style.display = '';
                } else {
                    document.getElementById('link' + other_sizes[i]).style.display = 'none';
                }
            } else {
                document.getElementById('link' + other_sizes[i]).href = 'image-' + other_sizes[i] + dbltilda_htmlsuffix + '#current=' + eval("elements_" + other_sizes[i] + "[current]");
            }
        }
        document.getElementById('thumbnails').href = 'thumbnails-' + dbltilda_current_size + '-' + thumbnailspages[current] + dbltilda_htmlsuffix + '#' + images[current];
        set_cursor("default");
    } else {
        setTimeout("show_current_text()", 50);
        set_cursor("wait");
    }
}

function display_current() {
    var main_img = document.getElementById('main_img');
    if (types[current] == 'image') {
        main_img.innerHTML = '<div class="fullscreen_image"><img src="' + images[current] + '"/></div>';
    } else {
        main_img.innerHTML = '<a class="fullscreenvideolink" href="' + videos[current] + '" '
                           + '   style="display:block;width:' + widths[current] + 'px;height:' + (heights[current] + 24) + 'px" id="player">'
                           + '  <div class="fullscreen_video"><img src="' + images[current] + '"/></div>'
                           + '  <img src="' + dbltilda_pathtobase + 'play_video.png" style="position:relative;top:-' + (heights[current] + 48)/2 + 'px;border:0;background-color:transparent"/>'
                           + '</a>';
        if (dbltilda_flowplayer_active == 'true') {
            flowplayer("player", dbltilda_pathtobase + "flowplayer-3.2.16.swf");
        }
    }
    oldhref = browser_href();
    newhref = 'image-' + dbltilda_current_size + dbltilda_htmlsuffix + '#current=' + images[current];
    if (oldhref != newhref) {
        currentURL = 'ignore1';
        location.href = newhref;
        currentURL = 'ignore2';
    }
    show_current_text();
    update_sensibilities();
}

function first() {
    if (slideshow == 1) {
        toggle_slideshow(true);
    }
    
    current = 0;
    display_current();
}

function next() {
    if (slideshow == 1) {
        toggle_slideshow(true);
    }

    if (current < images.length - 1) {
        current++;
        display_current();
    }
}

function next10() {
    if (slideshow == 1) {
        toggle_slideshow(true);
    }

    if (current < images.length - 11) {
        current += 10;
    } else {
        current = images.length - 1;
    }
    display_current();
}

function previous() {
    if (slideshow == 1) {
        toggle_slideshow(true);
    }

    if (current > 0) {
        current--;
        display_current();
    }
}

function previous10() {
    if (slideshow == 1) {
        toggle_slideshow(true);
    }

    if (current > 10) {
        current -= 10;
    } else {
        current = 0;
    }
    display_current();
}

function last() {
    if (slideshow == 1) {
        toggle_slideshow(true);
    }

    current = images.length - 1;
    display_current();
}

function toggle_video() {
    if (types[current] == 'video' && dbltilda_flowplayer_active == 'true') {
        var player = flowplayer('player');
        if (player.isLoaded()) {
            player.toggle();
        } else {
            player.play();
        }
    }
}

function keyDownEvent(key) {
    if (!key) {
        key = event;
        key.which = key.keyCode;
    }
    if (key.altKey || key.ctrlKey || key.shiftKey) {
        return;
    }
    switch (key.which) {
      case 32: // space
        toggle_video();
        break;
      case 36: // home
        first();
        break;
      case 35: // end
        last();
        break;
      case 37: // left
        previous();
        break;
      case 39: // right
        next();
        break;
      case 38: // up
        previous10();
        break;
      case 40: // down
        next10();
        break;
    }
}

function toggle_slideshow(now) {
    if (slideshow == 0) {
        slideshow_pause = document.getElementById('secs').value;
        add_cookie('booh-slideshow-pause-' + dbltilda_theme + '=' + slideshow_pause)
        document.getElementById("b_slideshow").value = dbltilda_stop_slideshow;
        slideshow = 1;
        if (current == images.length - 1) {
            current = -1;
        }
        if (now) {
            run_slideshow();
        } else {
            setTimeout("run_slideshow()", slideshow_pause * 1000);
        }
    } else {
        clearTimeout(slideshow_timer);
        document.getElementById("b_slideshow").value = dbltilda_run_slideshow;
        slideshow = 0;
    }
}

function run_slideshow() {
    if (slideshow == 0) {
        return;
    }

    if (images_loaded[current + 1] == 2) {
        current++;
        display_current();
        slideshow_timer = setTimeout("run_slideshow()", slideshow_pause * 1000);
    } else {
        slideshow_timer = setTimeout("run_slideshow()", 50);
    }

    if (current == images.length - 1) {
        toggle_slideshow(true);
    }
}

function set_preferred_size(val) {
    var expires = new Date(new Date().getTime() + (30 * 86400000));  // 30 days
    document.cookie = 'booh-preferred-size-' + dbltilda_theme + '='
                      + val
                      + '; expires=' + expires.toGMTString()
                      + '; path=/';
}

